### PR TITLE
*: support auto-compaction with finer granularity

### LINKED ISSUE #8503: support <1hr automatic compaction retention
Systems (e.g., k8s) that expect a lot of writes must implement their own compaction logic to periodically compact etcd at a rate of less than an hour. That is, some agent outside of etcd must be granted ownership of compaction which complicates system design. Since the intent is only etcd should compact more frequently, it would be much simpler to support automatic retention with finer granularity.

Currently the flag for configuring the compaction interval is an integer:
```
         --auto-compaction-retention '0'
                auto compaction retention length. 0 means disable auto compaction.
```

It could instead be parsed from a string:
```
        --auto-compaction-mode 'periodic'
                'periodic' means hours if an integer or a duration string otherwise, 'revision' means revision numbers to retain by auto compaction
```

Then, etcd can get 5 minute retention via `time.ParseDuration`:

`--auto-compaction-retention 5m`

Related #8458

### PR DESCRIPTION
**Pull request by fanminshi**
fixes #8503

### CODE REVIEW

**Comment by fanminshi** [comment]
Some manual test result:

test 5s interval:

```
$ bin/etcd --auto-compaction-retention 5s
2017-09-14 16:11:32.726296 N | compactor: Starting auto-compaction at revision 1 (retention: 5s )
2017-09-14 16:11:32.729183 I | mvcc: store.index: compact 1
2017-09-14 16:11:32.729237 N | compactor: Finished auto-compaction at revision 1
2017-09-14 16:11:32.729302 I | mvcc: finished scheduled compaction at 1 (took 41.163µs)
2017-09-14 16:11:37.726362 N | compactor: Starting auto-compaction at revision 1 (retention: 5s )
2017-09-14 16:11:37.726519 N | compactor: Finished auto-compaction at revision 1
2017-09-14 16:11:42.726478 N | compactor: Starting auto-compaction at revision 1 (retention: 5s )
2017-09-14 16:11:42.726642 N | compactor: Finished auto-compaction at revision 1
```

test no compaction:
```
$ bin/etcd --auto-compaction-retention 0
2017-09-14 16:12:08.019460 N | embed: serving insecure client requests on 127.0.0.1:2379, this is strongly discouraged!
```

test auto-compaction doesn't compact at every 1 `time.Duration(1)`:
```
$bin/etcd --auto-compaction-retention 1
2017-09-14 16:12:30.743382 N | embed: serving insecure client requests on 127.0.0.1:2379, this is strongly discouraged!
```

**Comment by fanminshi** [comment]
if the approach seems fine, I'll refactor the tests for compactor as it is currently tightly couple with  compactor's implementation.

File: compactor/periodic.go [hunk 1]
```diff
@@ -26,72 +26,60 @@ import (
...
-				plog.Noticef("Failed auto-compaction at revision %d (%v)", err, rev)
-				plog.Noticef("Retry after %v", checkCompactionInterval)
+				t.mu.RUnlock()
+				rev := t.rg.Rev()
+				plog.Noticef("Starting auto-compaction at revision %d (retention: %v )", rev, t.period)
+				_, err := t.c.Compact(t.ctx, &pb.CompactionRequest{Revision: rev})
+				if err == nil || err == mvcc.ErrCompacted {
+					plog.Noticef("Finished auto-compaction at revision %d", rev)
+				} else {
+					plog.Noticef("Failed auto-compaction at revision %d (%v)", err, rev)
```
Comment (gyuho) [inline compactor/periodic.go]: `, rev, err`?

Comment (fanminshi) [reply compactor/periodic.go]: good catch

File: compactor/periodic.go [hunk 2]
```diff
@@ -26,72 +26,60 @@ import (
...
+					t.mu.RUnlock()
 					continue
 				}
-			}
-
-			if clock.Now().Sub(last) < executeCompactionInterval {
-				continue
-			}
-
-			rev, remaining := t.getRev(t.periodInHour)
```
Comment (heyitsanthony) [inline compactor/periodic.go]: it's not necessary to remove this to support the feature and removing it completely changes the behavior of the periodic compactor

File: compactor/periodic.go [hunk 3]
```diff
@@ -26,72 +26,60 @@ import (
...
-			_, err := t.c.Compact(t.ctx, &pb.CompactionRequest{Revision: rev})
-			if err == nil || err == mvcc.ErrCompacted {
-				t.revs = remaining
-				last = clock.Now()
-				plog.Noticef("Finished auto-compaction at revision %d", rev)
-			} else {
-				plog.Noticef("Failed auto-compaction at revision %d (%v)", err, rev)
-				plog.Noticef("Retry after %v", checkCompactionInterval)
+				t.mu.RUnlock()
+				rev := t.rg.Rev()
```
Comment (xiang90) [inline compactor/periodic.go]: this is wrong. we should keep the previous logic to compact the revision of the past last x duration.

**Comment by fanminshi** [comment]
@xiang90 I changed the compactor logic with following:
1. get store rev now. 
2. wait for retention period to pass.
3. compact at rev from step 1.
4. go to step 1.

**Comment by xiang90** [comment]
why do we change the previous logic? i feel all we need to do there is to change the wait time from a fix x hours to some user given timeout, no?

**Comment by fanminshi** [comment]
@xiang90 the previous logic has a `checkCompactionInterval` that defaults to 5 mins. So the loop runs every 5 minutes to check if it should compact. It works for any wait time > 5 mins but not < 5  mins. therefore, I get rid of it. Having a looping checking every 5 minutes seems unnecessary. All the compactor really care is the rev at `time.Now() - retentionPeriod`. So I don't see the need to keep tracks revs every 5 minutes and then get the right rev through some math.

**Comment by xiang90** [comment]
@fanminshi 

There are a few issues with your approach, which the old approach addressed.

1. short retry time when error occurs (5 minutes instead of 1hr)
2. not updating compaction rev when error occurs

I would suggest you understand how the previous code handles failures before simplifying it. The rev recording part can be simplified though.

**Comment by fanminshi** [comment]
@xiang90 i see. let me rethink about the error handling part.

**Comment by fanminshi** [comment]
@xiang90 i add fast retry behavior into the code.

**Comment by xiang90** [comment]
this is getting more complicated. i am wondering why not keep the exact previous logic and change the checkInterval to min(user-specified-timeout/10, 1*second) or something.

**Comment by fanminshi** [comment]
@xiang90 checking every `checkCompactionInterval` to see if I should compact seems a bit heavy to me. I think there can be a more efficient way. Anyways, I can keep the original approach by changing `checkCompactionInterval`  to a appropriate value.

**Comment by xiang90** [comment]
> checking every checkCompactionInterval to see if I should compact seems a bit heavy to me. I think there can be a more efficient way. 

99% people will compact > 1 minute interval. that is max 0.16 read request / second, which is almost nothing.

**Comment by fanminshi** [comment]
@xiang90 redone the pr with previous approach. PTAL

File: compactor/periodic.go [hunk 4]
```diff
@@ -38,26 +38,30 @@ type Periodic struct {
...
-		c:            c,
+		clock:  clockwork.NewRealClock(),
+		period: h,
+		rg:     rg,
+		c:      c,
 	}
 }
 
+// N divides Periodic.period in into checkCompactionInterval duration
+var N = 10
```
Comment (xiang90) [inline compactor/periodic.go]: why not a const? why make this public?

Comment (xiang90) [reply compactor/periodic.go]: probably name this to something more descriptive too.

File: compactor/periodic.go [hunk 5]
```diff
@@ -66,31 +70,27 @@ func (t *Periodic) Run() {
 			case <-t.ctx.Done():
 				return
 			case <-clock.After(checkCompactionInterval):
-				t.mu.Lock()
+				t.mu.RLock()
```
Comment (xiang90) [inline compactor/periodic.go]: why change to use rlock here?

File: compactor/periodic.go [hunk 6]
```diff
@@ -113,8 +113,8 @@ func (t *Periodic) Resume() {
 	t.paused = false
 }
 
-func (t *Periodic) getRev(h int) (int64, []int64) {
-	i := len(t.revs) - int(time.Duration(h)*time.Hour/checkCompactionInterval)
+func (t *Periodic) getRev() (int64, []int64) {
```
Comment (xiang90) [inline compactor/periodic.go]: pass in n? probably we do not need the global defined const `n`.

**Comment by xiang90** [comment]
left some comments about minor issues. defer @gyuho or @jpbetz for a final look.

File: compactor/periodic.go (hunk 4 above)
Comment (fanminshi) [reply compactor/periodic.go]: yeah, i'll come up with a better name.

File: compactor/periodic.go (hunk 5 above)
Comment (fanminshi) [reply compactor/periodic.go]: i noticed that we are just reading `t.paused`. so I used rlock.

File: compactor/periodic.go (hunk 6 above)
Comment (fanminshi) [reply compactor/periodic.go]: the `n` can be used in testing. that's why I had a const `n`.

File: compactor/periodic.go (hunk 5 above)
Comment (xiang90) [reply compactor/periodic.go]: the critical area is super tight and not accessed frequently. it probably not worths changing it.

if you want to make it more efficient, do it in another PR. Let us keep one PR for one thing.

Comment (fanminshi) [reply compactor/periodic.go]: alright then

File: compactor/periodic.go (hunk 6 above)
Comment (xiang90) [reply compactor/periodic.go]: ah. ok. makes sense.

**Comment by gyuho** [comment]
@fanminshi Fix Travis CI failures?

File: etcdmain/config.go [hunk 7]
```diff
@@ -196,8 +196,8 @@ func newConfig() *config {
 	// version
 	fs.BoolVar(&cfg.printVersion, "version", false, "Print the version and exit.")
 
-	fs.IntVar(&cfg.AutoCompactionRetention, "auto-compaction-retention", 0, "Auto compaction retention for mvcc key value store. 0 means disable auto compaction.")
-	fs.StringVar(&cfg.AutoCompactionMode, "auto-compaction-mode", "periodic", "Interpret 'auto-compaction-retention' as hours when 'periodic', as revision numbers when 'revision'.")
+	fs.StringVar(&cfg.AutoCompactionRetention, "auto-compaction-retention", "0", "Auto compaction retention for mvcc key value store. 0 means disable auto compaction.")
+	fs.StringVar(&cfg.AutoCompactionMode, "auto-compaction-mode", "periodic", "'periodic' means hours if an integer or a duration string otherwise, 'revision' means revision numbers to retain by auto compaction")
```
Comment (jpbetz) [inline etcdmain/config.go]: I found the `Interpret 'auto-compaction-retention' as..` preamble helpful. It directly relates the two flags. Keep it?  Maybe:

`Interpret 'auto-compaction-retention' one of: periodic|revision. 'periodic' for duration based retention, defaulting to hours if no time unit is provided (e.g. '5m'). 'revision' for revision number based retention.` ?

File: compactor/periodic.go [hunk 8]
```diff
@@ -73,24 +77,20 @@ func (t *Periodic) Run() {
 					continue
 				}
 			}
-
-			if clock.Now().Sub(last) < executeCompactionInterval {
+			if clock.Now().Sub(last) < t.period {
```
Comment (jpbetz) [inline compactor/periodic.go]: Can we simplify the timer logic a bit here?  I found the logic a bit difficult to follow with the `case <-clock.After(checkCompactionInterval):` select case and the `periodDivisor` combined together.

Could we instead do something like 
```
select {
...
case <-clock.After(t.period - clock.Now().Sub(last)):
  ...
```

and then eliminate this `if clock.Now().Sub(last) < t.period {` check and the `periodDivisor`?  

(let me know if I'm missing an important subtlety to how the timers are being use here...)

File: compactor/periodic.go [hunk 9]
```diff
@@ -73,24 +77,20 @@ func (t *Periodic) Run() {
 					continue
 				}
 			}
-
-			if clock.Now().Sub(last) < executeCompactionInterval {
```
Comment (jpbetz) [inline compactor/periodic.go]: While we're improving this function. Can move the remaining code in the for loop into the `case <-clock.After` for readability?  It's only ever run as part of that `case`...

File: compactor/periodic.go (hunk 8 above)
Comment (fanminshi) [reply compactor/periodic.go]: @jpbetz I am trying to minimize any change. we can refactor the code in a future pr if you want.

File: compactor/periodic.go (hunk 9 above)
Comment (fanminshi) [reply compactor/periodic.go]: can we refactor the code in a future pr?

File: etcdmain/config.go (hunk 7 above)
Comment (fanminshi) [reply etcdmain/config.go]: your suggestion seems clearer. I'll change to that.

**Comment by fanminshi** [comment]
made the `auto-compaction-mode ` more clear.

**Comment by xiang90** [comment]
lgtm. defer to @gyuho 

@jpbetz let's improve the code in the next pr.

File: compactor/compactor.go [hunk 10]
```diff
@@ -29,8 +29,7 @@ var (
 )
 
 const (
-	checkCompactionInterval   = 5 * time.Minute
-	executeCompactionInterval = time.Hour
+	checkCompactionInterval = 5 * time.Minute
```
Comment (xiang90) [inline compactor/compactor.go]: this const can also be removed.

Comment (fanminshi) [reply compactor/compactor.go]: `checkCompactionInterval` is used `Revision` compactor. 
https://github.com/coreos/etcd/blob/master/compactor/revision.go#L64
I don't think it should be removed.

Comment (xiang90) [reply compactor/compactor.go]: ack

File: compactor/periodic.go [hunk 11]
```diff
@@ -38,60 +38,60 @@ type Periodic struct {
...
 			}
-
-			rev, remaining := t.getRev(t.periodInHour)
+			rev, remaining := t.getRev()
 			if rev < 0 {
 				continue
 			}
-
-			plog.Noticef("Starting auto-compaction at revision %d (retention: %d hours)", rev, t.periodInHour)
+			plog.Noticef("Starting auto-compaction at revision %d (retention: %v )", rev, t.period)
```
Comment (gyuho) [inline compactor/periodic.go]: `s/(retention: %v )/(retention: %v)/`

File: compactor/periodic.go (hunk 8 above)
Comment (jpbetz) [reply compactor/periodic.go]: Sounds good.

Comment (jpbetz) [reply compactor/periodic.go]: Sounds good.

File: compactor/periodic.go (hunk 11 above)
Comment (fanminshi) [reply compactor/periodic.go]: done.

File: etcdmain/help.go [hunk 12]
```diff
@@ -99,7 +99,7 @@ clustering flags:
 	--auto-compaction-retention '0'
 		auto compaction retention length. 0 means disable auto compaction.
 	--auto-compaction-mode 'periodic'
-		'periodic' means hours, 'revision' means revision numbers to retain by auto compaction
+		Interpret 'auto-compaction-retention' one of: periodic|revision. 'periodic' for duration based retention, defaulting to hours if no time unit is provided (e.g. '5m'). 'revision' for revision number based retention.
```
Comment (gyuho) [inline etcdmain/help.go]: `interpret` to make casing consistent?

Comment (fanminshi) [reply etcdmain/help.go]: sgtm

File: embed/etcd.go [hunk 13]
```diff
@@ -127,6 +128,24 @@ func StartEtcd(inCfg *Config) (e *Etcd, err error) {
 		}
 	}
 
+	var (
+		autoCompactionRetention time.Duration
+		h                       int
+	)
+	// AutoCompactionRetention defaults to "0" if not set.
+	if len(cfg.AutoCompactionRetention) == 0 {
```
Comment (fanminshi) [inline embed/etcd.go]: added a check see if AutoCompactionRetention is not set.

**Comment by fanminshi** [comment]
merge when green.

File: compactor/periodic.go [hunk 14]
```diff
@@ -38,60 +38,60 @@ type Periodic struct {
...
 			if rev < 0 {
 				continue
 			}
-
-			plog.Noticef("Starting auto-compaction at revision %d (retention: %d hours)", rev, t.periodInHour)
+			plog.Noticef("Starting auto-compaction at revision %d (retention: %v)", rev, t.period)
 			_, err := t.c.Compact(t.ctx, &pb.CompactionRequest{Revision: rev})
 			if err == nil || err == mvcc.ErrCompacted {
 				t.revs = remaining
-				last = clock.Now()
```
Comment (WIZARD-CXY) [inline compactor/periodic.go]: why we delete this line?

Comment (fanminshi) [reply compactor/periodic.go]: good a catch. I don't think this is meant to be deleted.

Comment (fanminshi) [reply compactor/periodic.go]: Actually, I dig into on why `last = clock.Now()` is deleted again. It is explained:

`I change the compaction frequency to every checkCompactionInterval when T > retentionPeriod.`
https://github.com/coreos/etcd/pull/8563#issuecomment-332364722

So the current behavior is not a bug but an behavior changed. I guess the question to ask is whether this intended behavior is correct or not.

**Pull request merged**